File: benchmark.js
/**
 * passByValue isnt actually possible in JS, so this is a load of BS. ignore this for now
 * in future - we may be able to distinguish properly, until then lets skip doing this in JS.
 *
 *
 * inside a code block in JS, let & consts are used as copies. whereas var is attached directly to parent.
 * maybe try to experiment and show something around this.
 */

const { performance } = require("perf_hooks");

function passByValue(arr) {
  let newArr = [...arr];
  // for (let i = 0; i < newArr.length; i++) {
  //   newArr[i] *= 2;
  // }
  return newArr;
}

function passByReference(arr) {
  for (let i = 0; i < arr.length; i++) {
    arr[i] *= 2;
  }
  return arr;
}

function benchmark(func, arr, iterations) {
  const start = performance.now();

  for (let i = 0; i < iterations; i++) {
    func([...arr]);
  }

  const end = performance.now();
  return end - start;
}

const testArray = Array(1000000).fill(1);
const iterations = 1000;

console.log("running PBV");
const valueTime = benchmark(passByValue, testArray, iterations);
console.log("running PBR");
const referenceTime = benchmark(passByReference, testArray, iterations);

// Print results
console.log(`Pass by value time: ${valueTime.toFixed(2)} ms`);
console.log(`Pass by reference time: ${referenceTime.toFixed(2)} ms`);
console.log(
  `Pass by reference is ${(valueTime / referenceTime).toFixed(2)}x faster`
);

File: outputs/benchmark.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <algorithm>

// Function using pass by value
std::vector<int> passByValue(std::vector<int> arr) {
    return arr;
}


// Function using pass by reference
std::vector<int> passByReference(std::vector<int>& arr) {
    return arr;
}

// Benchmark function
double benchmark(auto func, const std::vector<int>& arr, int iterations) {
    auto start = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < iterations; ++i) {
        std::vector<int> copy = arr; // Create a new copy of the array for each iteration
        func(copy);
    }

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double, std::milli> duration = end - start;
    return duration.count();
}

int main() {
    // Create a large vector for testing
    std::vector<int> testVector(1000000, 1);
    int iterations = 1000;

    // Run benchmarks
    double valueTime = benchmark(passByValue, testVector, iterations);
    double referenceTime = benchmark([](std::vector<int>& v) { passByReference(v); }, testVector, iterations);

    // Print results
    std::cout << "Pass by value time: " << valueTime << " ms\n";
    std::cout << "Pass by reference time: " << referenceTime << " ms\n";
    std::cout << "Pass by reference is " << valueTime / referenceTime << "x faster\n";

    return 0;
}
/*
Program Output:

===============
Pass by value time: 2539.23 ms
Pass by reference time: 2487.34 ms
Pass by reference is 1.02086x faster
*/
======================================================================================================================================================
